ksql-api/src/main/java/io/confluent/ksql/api/plugin/KsqlServerEndpoints.java

```diff
@@ -0,0 +1,210 @@
+ */
+
+package io.confluent.ksql.api.plugin;
+
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import io.confluent.ksql.api.impl.Utils;
+import io.confluent.ksql.api.server.BaseServerEndpoints;
+import io.confluent.ksql.api.server.PushQueryHandler;
+import io.confluent.ksql.api.spi.InsertsSubscriber;
+import io.confluent.ksql.engine.KsqlEngine;
+import io.confluent.ksql.parser.KsqlParser.ParsedStatement;
+import io.confluent.ksql.parser.KsqlParser.PreparedStatement;
+import io.confluent.ksql.parser.tree.Query;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.query.RowConsumer;
+import io.confluent.ksql.schema.ksql.Column;
+import io.confluent.ksql.schema.ksql.FormatOptions;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.registry.KsqlSchemaRegistryClientFactory;
+import io.confluent.ksql.security.KsqlSecurityExtension;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.statement.ConfiguredStatement;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlStatementException;
+import io.confluent.ksql.util.QueryMetadata;
+import io.vertx.core.Context;
+import io.vertx.core.WorkerExecutor;
+import io.vertx.core.json.JsonObject;
+import java.security.Principal;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.function.Supplier;
+import org.apache.kafka.streams.KafkaClientSupplier;
+import org.apache.kafka.streams.processor.internals.DefaultKafkaClientSupplier;
+import org.reactivestreams.Subscriber;
+
+public class KsqlServerEndpoints extends BaseServerEndpoints {
+
+  private final KsqlEngine ksqlEngine;
+  private final KsqlConfig ksqlConfig;
+  private final KsqlSecurityExtension securityExtension;
+  private final ServiceContextFactory theServiceContextFactory;
+
+  public interface ServiceContextFactory {
+
+    ServiceContext create(
+        KsqlConfig ksqlConfig,
+        Optional<String> authHeader,
+        KafkaClientSupplier kafkaClientSupplier,
+        Supplier<SchemaRegistryClient> srClientFactory
+    );
+  }
+
+  public KsqlServerEndpoints(
+      final KsqlEngine ksqlEngine,
+      final KsqlConfig ksqlConfig,
+      final KsqlSecurityExtension securityExtension,
+      final ServiceContextFactory theServiceContextFactory) {
+    this.ksqlEngine = Objects.requireNonNull(ksqlEngine);
+    this.ksqlConfig = Objects.requireNonNull(ksqlConfig);
+    this.securityExtension = Objects.requireNonNull(securityExtension);
+    this.theServiceContextFactory = Objects.requireNonNull(theServiceContextFactory);
+  }
+
+  @Override
+  protected PushQueryHandler createQuery(final String sql, final JsonObject properties,
+      final Context context, final WorkerExecutor workerExecutor, final RowConsumer rowConsumer) {
+    // Must be run on worker as all this stuff is slow
```

Comment:
Pardon the ignorance, but how do we know "this stuff is slow"?

Reply:
Anything taking more than a very small number of milliseconds will be slow for an event loop. I think we should err on the side of caution.